### PR TITLE
Refactor serve_packet into IO and non IO parts

### PR DESCRIPTION
**Pull request by tdittr**
This decouples the IO during packet handling from the actual logic. This allows easier writing of tests,

This PR also flattens a lot of nested matches to make them more readable.

### CODE REVIEW

File: ntpd/src/daemon/server.rs [hunk 1]
```diff
@@ -196,321 +283,262 @@ impl<C: 'static + NtpClock + Send> ServerTask<C> {
...
-                };
+        }
+    }
 
-                if let Err(serialize_err) = serialize_result {
-                    self.stats.response_send_errors.inc();
-                    warn!(error=?serialize_err, "Could not serialize response");
-                    return true;
-                }
+    // TODO: this instrument breaks coverage collection for some reason...
```
Comment (davidv1992) [inline ntpd/src/daemon/server.rs:367]: Remaining todo

File: ntpd/src/daemon/server.rs [hunk 2]
```diff
@@ -196,321 +283,262 @@ impl<C: 'static + NtpClock + Send> ServerTask<C> {
...
-                    return true;
                 }
+            }
+            Ok((length, peer_addr, opt_timestamp)) => {
+                let Some(request_buf) = buf.get(..length) else {
+                    warn!("length from socket is out of bounds. This is a bug!");
+                    return SocketConnection::Reconnect;
+                };
+                let mut response_buf = [0; MAX_PACKET_SIZE];
+                let response_buf = response_buf.as_mut_slice();
```
Comment (davidv1992) [inline ntpd/src/daemon/server.rs:345]: We should perhaps limit the response buffer size here already.

**Review by davidv1992 — COMMENTED** [review]
Two minor points, but otherwise looking very good.

File: ntpd/src/daemon/server.rs (hunk 1 above)
Comment (tdittr) [reply ntpd/src/daemon/server.rs]: Fixed!

File: ntpd/src/daemon/server.rs (hunk 2 above)
Comment (tdittr) [reply ntpd/src/daemon/server.rs]: Great idea, changed it and removed the old check at the end. So now if this ever happens it should just show up as a serialization error.

**Review by davidv1992 — APPROVED** [review]

**Pull request merged**